### DISCUSSION TITLE
Automatic vs Manual Layouting

### DISCUSSION THREAD

**Discussion by jmenzies12** [eclipse-glsp/glsp#1414] [Q&A] [answered]
Hey there,

I'm currently trying to implement layouting in my diagram. When I trigger the layout manually, elk returns the changes but they don't get applied. When I have the layout on automatic, it does actually apply the changes. I assumed that if automatic worked then manual would too, am I missing something?

Using GLSP 2.2.1 server-node and vscode-integration.

Kind Regards
Jonathan

**Comment by tortmayr** [comment, marked as answer]
Hi @jmenzies12,

unfortunatley this is a known issue when using manual layout (https://github.com/eclipse-glsp/glsp/issues/1173).
The implementation of the `LayoutOperationHandler` is faulty and does not work generically for custom source models.
Currently the `ELkLayoutEngine` directly manipulates the graphical model (GModel). This works fine for the automatic layout use case
but does not work for the manual use case. 
When automatic layout is used the layout information does not need to be persisted in the source model because it will be recalculated after every model update anyways. For manual layouting this is not true and the layout changes (size and position of elements) needs to be stored in the source model. This step is currently missing in the base framework. That`s why manual layout is not working.

Until this is fixed you could rebind the default `LayoutOperationHandler` to a custom implementation that updates your source model accordingly after the layout engine has been extecuted.

**Reply by jmenzies12** [reply]
Oh right, I see what you mean. That is unfortunate but thank you for the workaround suggestion, I'll go down that route for now then.

**Reply by jmenzies12** [reply]
Hey, just wanted to comment and say that the workaround ended up being fairly straightforward and I now have manual layouting working.
